### ISSUE TITLE
EventEmitter Does Not Emit Inside a Promise

### ISSUE THREAD

**Issue by mouhssine-ka** [angular/angular#58723]
### Which @angular/* package(s) are the source of the bug?

zone.js, Don't known / other

### Is this a regression?

No

### Description

When emitting an event using EventEmitter inside a promise in a child component, the event does not propagate correctly to the parent. This prevents the parent from receiving the emitted event and handling it, effectively blocking the flow of communication between components.

Steps to Reproduce:
Use app.component.ts as a child component that emits an event inside a promise using EventEmitter.
crate a parent component, and listen to the emitted event and process it (e.g., resolve or reject the promise).
Observe that the parent never receives the event, and the subsequent logic in the child component is not executed.

Example Code:

Child Component (TS):
```
import { Component, EventEmitter, Output } from '@angular/core';

@Component({
  selector: 'app-root',
  standalone: true,
  templateUrl: './app.component.html',
  styleUrl: './app.component.css',
})
export class AppComponent {
  title = 'test';
  count: number = 0;

  @Output() counterIncrementedEventEvent = new EventEmitter<any>();

  protected async check() {

    await new Promise<any>((resolve, reject) => {
      console.log("here reach")
      this.counterIncrementedEventEvent.emit({ resolve, reject });
    });

    console.log("here doesnt reach")
    this.count++;
  }
}

```

Parent Component (HTML):
`<app-root (counterIncrementedEventEvent)="counterIncrementedEvent($event)"></app-root>`

Parent Component (HTML): 
```
import { Component } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { AppComponent } from '../app.component';

@Component({
  selector: 'app-parent',
  standalone: true,
  imports: [AppComponent],
  templateUrl: './parent.component.html',
  styleUrls: ['./parent.component.css'],
})
export class ParentComponent {
  constructor(private httpClient: HttpClient) {}

  async counterIncrementedEvent($event: any) {
    console.log("Handling event in ParentComponent... HERE DOESNT REACH");
    const { resolve, reject } = $event;

    // Example logic: Simulate a network call
    this.delete().subscribe({
      next: () => {
        resolve(); 
      },
      error: () => {
        reject(); 
      },
    });
  }

  delete() {
    return this.httpClient.delete("https://jsonplaceholder.typicode.com/todo/1");
  }
}

```

### Please provide the exception or error you saw

```true
No exception or error is thrown during this process
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 18.2.10
Node: 20.14.0
Package Manager: npm 10.8.2
OS: win32 x64

Angular: 18.2.9
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1802.10
@angular-devkit/build-angular   18.2.10
@angular-devkit/core            18.2.10
@angular-devkit/schematics      18.2.10
@angular/cli                    18.2.10
@schematics/angular             18.2.10
rxjs                            7.8.1
typescript                      5.4.5
zone.js                         0.14.10
```

**Comment by JeanMeche** [comment]
Hi, could you please provide a working reproduction (for example via stackblitz) for us to have a look. Thank you.

**Comment by mouhssine-ka** [comment]
> Hi, could you please provide a working reproduction (for example via stackblitz) for us to have a look. Thank you.

https://stackblitz.com/edit/angular-vakejq-a9s2cg?file=src%2Fapp.component.ts

for more context, open the console

**Comment by mouhssine-ka** [comment]
https://stackblitz.com/edit/angular-vakejq-a9s2cg?file=src%2Fapp.component.ts

for more context, open the console

Il giorno lun 18 nov 2024 alle ore 20:41 Matthieu Riegler <
***@***.***> ha scritto:

**Comment by JeanMeche** [comment]
Fyi, zone based apps are broken on the old stackblitz engine if the ts target is >= es2017. Native async/await are not leveled down and they aren't patched by zone.js. 

Do you reproduce this issue outside of stackblitz ?

**Comment by mouhssine-ka** [comment]
Yes, the issue occurs outside of StackBlitz as well. To reproduce it, simply create a new project and use the code provided in my original report.

**Comment by JeanMeche** [comment]
Actually in your example the issue is that your promise is never resolved, the `await` waits indefinitely.

**Comment by mouhssine-ka** [comment]
As I mentioned earlier, _the parent never receives the event, and the subsequent logic in the child component is not executed_. 

**The parent component is the one responsible for calling the resolve() method on the promise.** 

Here's how it works: When the event is emitted using `this.counterIncrementedEvent.emit({ resolve, reject });` It sends references to the resolve and reject functions. The idea is that when the parent receives the event and calls resolve(), the promise should be resolved, and the child component logic proceeds as expected. On the other hand, if the parent calls reject(), an error is thrown (which is what I have used in different contexts). In those cases, I catch the error using a try-catch block

In the simplified code I provided, I removed unnecessary parts for clarity.

In other words, the parent is not calling resolve() because it never receives the emitted event. This prevents the promise from being resolved and halts the logic flow in the child component.

**Comment by JeanMeche** [comment]
Again I'm not able to reproduce your issue. 
You stackblitz example doesn't exhibit the problem. 

This is what is working for me : 

```
import { bootstrapApplication } from '@angular/platform-browser';
import { appConfig } from './app/app.config';
import { Component, EventEmitter, Output } from '@angular/core';

@Component({
  selector: 'app-child',
  standalone: true,
  template: `<h1>Title</h1>

  <button (click)="check()">click to increment and check if count is greater than 0</button>
  
  <p>Count: {{count}}</p>
  
  @if(count > 0){
    count is greater than 0
  }
  @else {
    count is NOT greater than 0
  }`,
})
export class ChildComponent {
  title = 'test';
  count: number = 0;

  @Output() counterIncrementedEventEvent = new EventEmitter<any>();

  protected async check() {
    await new Promise<any>((resolve, reject) => {
      console.log('here reach');
      this.counterIncrementedEventEvent.emit({ resolve, reject });
      console.log('here reach 2');
    });

    console.log('here doesnt reach');
    this.count++;
  }
}


@Component({
  selector: 'app-root',
  template: `<app-child (counterIncrementedEventEvent)="onCounterIncrementedEvent($event)"></app-child>`,
  imports: [ChildComponent],
  standalone: true,
})
export class AppComponent {
  onCounterIncrementedEvent(e:any) {
    e.resolve();
    console.log('ok')
  }
}

bootstrapApplication(AppComponent, appConfig)
  .catch((err) => console.error(err));
```
